### PR TITLE
Split cleaner cycle for active and deleted tenants

### PR DESCRIPTION
**Pull request by alexqyle**
**What this PR does**:

Split cleaner cycle into two separated cycles. One is for active tenants to be cleaned for deleted blocks and update bucket index. Another cycle is for deleted tenants to be cleaned for any data upon deletion. The reason separating those two is that sometime cleaning data for deleted tenant could take a long time to fully delete all blocks if the tenant is large enough. This would block the rest of tenants to be cleaned or deleted which would result in delay bucket index update for active tenants. Separating could prevent this kind of blockage to happen

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/compactor/blocks_cleaner.go [hunk 1]
```diff
@@ -129,54 +137,175 @@ func NewBlocksCleaner(
...
 	// Run a cleanup so that any other service depending on this service
 	// is guaranteed to start once the initial cleanup has been done.
-	c.runCleanup(ctx, true)
+	activeUsers, deletedUsers, err := c.scanUsers(ctx)
 
+	if err != nil {
+		level.Error(c.logger).Log("msg", "failed to scan users on startup", "err", err.Error())
+		c.runsFailed.WithLabelValues(deletedStatus).Inc()
+		c.runsFailed.WithLabelValues(activeStatus).Inc()
+		return nil
```
Comment (yeya24) [inline pkg/compactor/blocks_cleaner.go:173]: The reason why we don't return error is for cleaner to retry at `loop`?
This looks weird. I don't know if other Cortex components do the same thing

Comment (alexqyle) [reply pkg/compactor/blocks_cleaner.go:173]: It is not intended to return `nil`. Let me revisit the code and update it.

Comment (alexqyle) [reply pkg/compactor/blocks_cleaner.go:173]: Based on current logic (https://github.com/cortexproject/cortex/blob/master/pkg/compactor/blocks_cleaner.go#L144), cleaner never returns error. If it failed, it would just wait for next tick and run it again.

File: pkg/compactor/blocks_cleaner.go [hunk 2]
```diff
@@ -129,54 +137,175 @@ func NewBlocksCleaner(
...
+	}
+}
+
+func (c *BlocksCleaner) runActiveUserCleanup(ctx context.Context, jobChan chan *cleanerJob) {
+	for job := range jobChan {
+		if job.timestamp < time.Now().Add(-c.cfg.CleanupInterval).Unix() {
+			level.Warn(c.logger).Log("Active user cleaner job too old. Ignoring to get recent data")
+			continue
+		}
+		c.cleanUpActiveUsers(ctx, job.users, false) //nolint:errcheck
```
Comment (danielblando) [inline pkg/compactor/blocks_cleaner.go:232]: Don't we need to check error here and increase metric if failed? Similar to what we do in `starting`?

File: pkg/compactor/blocks_cleaner.go [hunk 3]
```diff
@@ -129,54 +137,175 @@ func NewBlocksCleaner(
...
+		level.Info(c.logger).Log("msg", "successfully completed blocks cleanup and maintenance for active users")
+		c.runsCompleted.WithLabelValues(activeStatus).Inc()
+		c.runsLastSuccess.WithLabelValues(activeStatus).SetToCurrentTime()
 	} else if errors.Is(err, context.Canceled) {
-		level.Info(c.logger).Log("msg", "canceled blocks cleanup and maintenance", "err", err)
-		return
+		level.Info(c.logger).Log("msg", "canceled blocks cleanup and maintenance for active users", "err", err)
+	} else {
+		level.Error(c.logger).Log("msg", "failed to run blocks cleanup and maintenance for active users", "err", err.Error())
+		c.runsFailed.WithLabelValues(activeStatus).Inc()
```
Comment (danielblando) [inline pkg/compactor/blocks_cleaner.go:258]: Hmmm for `starting` we are check return and also increase the metric. For the `loop`, we are not doing anything, which then this seems correct. I think we need to make it the same in both places, or increase inline or check failure and increase on caller

File: pkg/compactor/blocks_cleaner.go [hunk 4]
```diff
@@ -129,54 +137,175 @@ func NewBlocksCleaner(
...
+		level.Info(c.logger).Log("msg", "successfully completed blocks cleanup and maintenance for deleted users")
+		c.runsCompleted.WithLabelValues(deletedStatus).Inc()
+		c.runsLastSuccess.WithLabelValues(deletedStatus).SetToCurrentTime()
+	} else if errors.Is(err, context.Canceled) {
+		level.Info(c.logger).Log("msg", "canceled blocks cleanup and maintenance for deleted users", "err", err)
 	} else {
-		level.Error(c.logger).Log("msg", "failed to run blocks cleanup and maintenance", "err", err.Error())
-		c.runsFailed.Inc()
+		level.Error(c.logger).Log("msg", "failed to run blocks cleanup and maintenance for deleted users", "err", err.Error())
+		c.runsFailed.WithLabelValues(deletedStatus).Inc()
```
Comment (danielblando) [inline pkg/compactor/blocks_cleaner.go:295]: same here

File: pkg/compactor/blocks_cleaner.go [hunk 5]
```diff
@@ -47,14 +50,16 @@ type BlocksCleaner struct {
 	bucketClient objstore.InstrumentedBucket
 	usersScanner *cortex_tsdb.UsersScanner
 
+	ringLifecyclerID string
```
Comment (danielblando) [inline pkg/compactor/blocks_cleaner.go:53]: Is this related to this PR? I cant find where it is being used. Only we setting it. Maybe it was mixed in the split of PRs

File: pkg/compactor/blocks_cleaner.go (hunk 2 above)
Comment (alexqyle) [reply pkg/compactor/blocks_cleaner.go]: Addresed

File: pkg/compactor/blocks_cleaner.go (hunk 3 above)
Comment (alexqyle) [reply pkg/compactor/blocks_cleaner.go]: Addresed

File: pkg/compactor/blocks_cleaner.go (hunk 4 above)
Comment (alexqyle) [reply pkg/compactor/blocks_cleaner.go]: Addresed

File: pkg/compactor/blocks_cleaner.go (hunk 5 above)
Comment (alexqyle) [reply pkg/compactor/blocks_cleaner.go]: Removed

File: pkg/compactor/blocks_cleaner.go (hunk 1 above)
Comment (yeya24) [reply pkg/compactor/blocks_cleaner.go:166]: This behavior sounds weird tbh. But let's leave it as it is. I guess if there is any error connecting to object storage like credential issues, it expects compactor to fail from compaction, not from cleaner.

**Review by yeya24 — APPROVED** [review]

**Review by danielblando — APPROVED** [review]
